File: sortOfFun/sorts.py
#bubble sort
def bubble_sort(arr):
  sorted = False
  while not sorted: #start bubbling
    sorted = True #sort until there are no more swaps
    for i, el in enumerate(arr):
      #avoid bounds error with this check
      if i < len(arr) - 1:
        if el > arr[i+1]:
          swap(arr, i, i+1)
          sorted = False
  return arr

#helper function for bubble sort
def swap(arr, i, j):
  temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
  return arr

#insertion sort
def insertion_sort(arr):
  insertion_idx = 0
  while insertion_idx < len(arr):
    insertion_el = arr[insertion_idx]
    for i, el in enumerate(arr[:insertion_idx]): #loop through every el within the range of insertion_idx
      if i + 1 >= insertion_idx: #compare el to the elements within the range of insertion_idx
        if arr[insertion_idx] < el:
          #remove insertion el and place it where it needs to go
          del arr[insertion_idx]
          arr.insert(i, insertion_el)
          #you don't need to check anymore elements
          break
    insertion_idx += 1
  return arr
          

#merge sort
def merge_sort(arr):
  #base case, the arr is sorted if it is of length 1 or 0
  if len(arr) < 2: 
    return arr
  
  #divide the array in two
  mid_idx = len(arr) // 2
  left_arr = arr[: mid_idx]
  right_arr = arr[mid_idx :]

  #recursively sort the two arrays
  sorted_left_arr = merge_sort(left_arr)
  sorted_right_arr = merge_sort(right_arr)

  #use helper function to merge the two sorted arrays
  return merge(sorted_left_arr, sorted_right_arr)

#helper function for merge sort
def merge(arr1, arr2):
  merged = []
  i = 0 #init indices
  j = 0
  while len(merged) < len(arr1) + len(arr2):
    #if either array has been completely checked, add the other array to merged
    if i >= len(arr1):
      merged.extend(arr2[j:])
      return merged
    elif j >= len(arr2):
      merged.extend(arr1[i:])
      return merged

    #if neither array is empty, check which element is smaller and add to merged
    if arr1[i] > arr2[j]:
      merged.append(arr2[j])
      j+=1
    else:
      merged.append(arr1[i])
      i+=1

  return merged

print("bubblesort: {bubble}".format(bubble = bubble_sort([4,2,3,5,78,9,3,21,1,80])))
print("insertion sort: {insertion} ".format(insertion = insertion_sort([4,2,3,5,78,9,3,21,1,80])))
print("merge sort: {merge}".format(merge = merge_sort([4,2,3,5,78,9,3,21,1,80])))

File: sortOfFun/sorts.c
#include <stdio.h>
#include <stdbool.h> //include true and false

// HELPER FUNCTIONS 
void print_array(int *arr, int length) {
    // bc arrays are passed as pointers
    // you need the length of the array to print the entire array
    int i = 0;
    for (i = 0; i < length; i++) {
        printf("%d, ", arr[i]);
    }
    printf("\n");
}

int * swap( int *arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
    return arr;
}

int * slice_array( int *arr, int length_arr,  int start, int finish) {
    //slice array is inclusive
    int sliced[finish-start];
    int slice_i = 0;
    int i;
    for (i = start; i <= finish; i++ ) {
        sliced[slice_i] = arr[i];
        slice_i += 1;
    }
    printf("sliced array: ");
    print_array(sliced, finish - start);
    return sliced;
}

// SORTING ALGORITHMS

// bubble sort
int * bubble_sort( int *arr, int length) {
    bool sorted = false;
    int i = 0;
    while (!sorted) {
        sorted = true;
        for (i = 0; i < length; i++) {
            if (i + 1 < length) { //only check for swap up to the second to last element
                if (arr[i] > arr[i+1]) {
                    swap(arr, i, i+1);
                    sorted = false;
                }
            }
        }
    }
    return arr;
}
// insertion sort

// merge sort

void merge(int arr[], int leftIndex, int midIndex, int rightIndex) {
    int i,j,k;
    int arr1Len = midIndex - leftIndex +1;
    int arr2Len = rightIndex - midIndex;

    int arr1[arr1Len], arr2[arr2Len];
    //copy data to the temp arrays: arr1 and arr2
    for (i = 0; i < arr1Len; i++) {
        arr1[i] = arr[leftIndex + i];
    }
    for (j = 0; j < arr2Len; j++) {
        arr2[j] = arr[midIndex + 1 + j];
    }

    i = 0;
    j = 0;
    k = leftIndex;

    //start merging the arrays
    while (i < arr1Len && j < arr2Len) {
        if (arr1[i] <= arr2[j]) {
            arr[k] = arr1[i];
            i++;
        } else {
            arr[k] = arr2[j];
            j++;
        }
        k++;
    }

    // copy the remaining elements of arr1 if there are any
    while (i < arr1Len) {
        arr[k] = arr1[i];
        i++;
        k++;
    }

    // copy the remaining elements of arr2 if there are any
    while (j < arr2Len) {
        arr[k] = arr2[j];
        j++;
        k++;
    }
}
// modifies the incoming array
// https://www.geeksforgeeks.org/c-program-for-merge-sort/
void merge_sort(int arr[], int leftIndex, int rightIndex) {
    if (leftIndex < rightIndex) { //do some sorting
        int midIndex = (leftIndex + (rightIndex)) / 2;

        merge_sort(arr, leftIndex, midIndex);
        merge_sort(arr, midIndex + 1, rightIndex);
        merge(arr, leftIndex, midIndex, rightIndex);
    }
}

int main() {
    int sortThisArray[] = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    int arrSize = sizeof(sortThisArray)/sizeof(int);
    int * sorted = bubble_sort(sortThisArray, arrSize);
    printf("bubble_sort: ");
    print_array(sorted, arrSize);

    int sortThisArrayWithMergeSort[] = {15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    int arrSizeMerge = sizeof(sortThisArrayWithMergeSort)/sizeof(int);
    printf("The length of the array we are passing into merge sort is: %d \n", arrSizeMerge - 1);
    merge_sort(sortThisArrayWithMergeSort, 0, arrSizeMerge - 1);

    printf("merge_sort: ");
    print_array(sortThisArrayWithMergeSort, arrSizeMerge);
}